pkg/controller/logstorage/logstorage.go

```diff
@@ -0,0 +1,288 @@
+package logstorage
+
+import (
+	"context"
+	"fmt"
+
+	kbv1 "github.com/elastic/cloud-on-k8s/pkg/apis/kibana/v1"
+	apps "k8s.io/api/apps/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	cmnv1 "github.com/elastic/cloud-on-k8s/pkg/apis/common/v1"
+	esv1 "github.com/elastic/cloud-on-k8s/pkg/apis/elasticsearch/v1"
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	storagev1 "k8s.io/api/storage/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	operatorv1 "github.com/tigera/operator/api/v1"
+	"github.com/tigera/operator/pkg/controller/utils"
+	"github.com/tigera/operator/pkg/controller/utils/imageset"
+	"github.com/tigera/operator/pkg/render"
+	relasticsearch "github.com/tigera/operator/pkg/render/common/elasticsearch"
+	rmeta "github.com/tigera/operator/pkg/render/common/meta"
+	rsecret "github.com/tigera/operator/pkg/render/common/secret"
+)
+
+func (r *ReconcileLogStorage) createLogStorage(
```

Comment:
I think all of these createXXX functions need to return a Reconcile object and an error, and return that to the caller of the controller. I say this because there are 3 types of errors:
1. A real error happened that we weren't expecting, and we expect an immediate retry to fix it (hopefully)
2. A temporary error, like something not being ready, and we want to retry in a specific amount of time
3. A temporary error, like a k8s resource not being ready, and we want to wait for a notification from the k8s API telling us that the state of a resource has changed

In order, the logging / what should be returned to the caller of the controller is as follows:
1. Log the error, set degraded, return an empty reconcile and an error
2. Log the error, set degraded, return a reconcile with `Requeue` or `RequeueAfter` set and a nil error
3. Log the error, set degraded, return an empty Reconcile and a nil error (we'll re reconcile when k8s notifies us of a change)

I commented on one the tests regarded 3, it's important we don't return an error otherwise we end up in a tight reconcile loop when we know there's nothing we can do until k8s notifies us about a change.

Reply:
One other thing, I think the createXXX functions need to either return a 3rd parameter saying we should continue, or return a special error that we can switch on. The caller wouldn't be able to tell if Reconcile is empty and the error is nil because everything is OK or because of the 3rd case above.

Reply:
I've asked @tmjd about his preference about returning a special error or a bool, as this is likely setting a precedent.

Reply:
For now I am checking to see if `Requeue` is set on the Reconcile object. I can add the third return value instead if that's what we end up going with.